src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java

```diff
@@ -0,0 +1,177 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.xml.security.encryption.keys.content.derivedKey;
+
+import org.apache.xml.security.encryption.XMLCipherUtil;
+import org.apache.xml.security.exceptions.XMLSecurityException;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+import java.nio.ByteBuffer;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+
+import static java.lang.System.Logger.Level.DEBUG;
+
+/**
+ * The implementation of the HMAC-based Extract-and-Expand Key Derivation Function (HKDF) as defined in RFC 5869.
+ * <p>
+ * The HKDF algorithm is defined as follows:
+ * <pre>
+ * N = ceil(L/HashLen)
+ * T = T(1) | T(2) | T(3) | ... | T(N)
+ * OKM = first L bytes of T
+ * where:
+ * T(0) = empty string (zero length)
+ * T(1) = HMAC-Hash(PRK, T(0) | info | 0x01)
+ * T(2) = HMAC-Hash(PRK, T(1) | info | 0x02)
+ * T(3) = HMAC-Hash(PRK, T(2) | info | 0x03)
+ * ...
+ * </pre>
+ */
+public class HKDF implements DerivationAlgorithm {
+
+    private static final System.Logger LOG = System.getLogger(HKDF.class.getName());
+    private final String hmacHashAlgorithmURI;
+    private final Mac hmac;
+
+    /**
+     * Constructor HKDF initializes the Mac object with the given algorithmURI and salt.
+     *
+     * @param hmacHashAlgorithmURI the Hash algorithm
+     * @param salt               the salt value to initialize the MAC algorithm.
+     * @throws XMLSecurityException if the key derivation initialization fails for any reason
+     */
+    public HKDF(String hmacHashAlgorithmURI, byte[] salt) throws XMLSecurityException {
```

Comment:
Why is salt specified here but not in the deriveKey method? This means this HKDF object cannot be reused if salt is different.

I think part of the problem is that the `DerivationAlgorithm` interface is too specific and designed to support ConcatKDF. It may make sense to add some more `deriveKey` methods to that interface that take a `KDFParams` object so that the parameters can be more abstract and defined by the subclasses, ex:

`byte[] deriveKey(byte[] secret, KDFParams params) throws XMLSecurityException`

Reply:
Done